trillian/ctfe/services.go

```diff
@@ -82,6 +105,115 @@ func (s *issuanceChainService) Add(ctx context.Context, chain []byte) ([]byte, e
 	return hash, nil
 }
 
+// BuildLogLeaf builds the MerkleTreeLeaf that gets sent to the backend, and make a trillian.LogLeaf for it.
+func (s *issuanceChainService) BuildLogLeaf(ctx context.Context, chain []*x509.Certificate, logPrefix string, merkleLeaf *ct.MerkleTreeLeaf, isPrecert bool) (*trillian.LogLeaf, error) {
+	raw := extractRawCerts(chain)
+
+	// If CTFE storage is enabled for issuance chain, add the chain to storage
+	// and cache, and then build log leaf. If Trillian gRPC is enabled for
+	// issuance chain, build the log leaf.
+	if s.IsCTFEStorageEnabled() {
+		issuanceChain, err := asn1.Marshal(raw[1:])
+		if err != nil {
+			return &trillian.LogLeaf{}, fmt.Errorf("failed to marshal issuance chain: %s", err)
+		}
+		hash, err := s.add(ctx, issuanceChain)
+		if err != nil {
+			return &trillian.LogLeaf{}, fmt.Errorf("failed to add issuance chain into CTFE storage: %s", err)
+		}
+		leaf, err := util.BuildLogLeafWithHash(logPrefix, *merkleLeaf, 0, raw[0], nil, hash, isPrecert)
+		if err != nil {
+			return &trillian.LogLeaf{}, fmt.Errorf("failed to build LogLeaf: %s", err)
+		}
+		return &leaf, nil
+	} else {
+		leaf, err := util.BuildLogLeaf(logPrefix, *merkleLeaf, 0, raw[0], raw[1:], isPrecert)
+		if err != nil {
+			return &trillian.LogLeaf{}, fmt.Errorf("failed to build LogLeaf: %s", err)
+		}
+		return &leaf, nil
+	}
+}
+
+// FixLogLeaf recreates the LogLeaf.ExtraData if CTFE storage backend is
+// enabled and the type of LogLeaf.ExtraData contains any hash (e.g.
+// PrecertChainEntryHash, CertificateChainHash).
+func (s *issuanceChainService) FixLogLeaf(ctx context.Context, leaf *trillian.LogLeaf) error {
+	// Skip if CTFE storage backend is not enabled.
+	if !s.IsCTFEStorageEnabled() {
+		return nil
+	}
+
+	var precertChainHash ct.PrecertChainEntryHash
```

Comment:
Might be worth a comment in here explaining that we don't have a way to know ahead of time which struct is stored in `ExtraData`, and so all we can do it to try to unmarshal it into each of the types in sequence?

Reply:
Done, thanks!